File: server/java/codegen-spring/src/main/java/com/metaobjects/generator/spring/ExtractSchemaEmitter.java
package com.metaobjects.generator.spring;

import com.metaobjects.field.BooleanField;
import com.metaobjects.field.DoubleField;
import com.metaobjects.field.EnumField;
import com.metaobjects.field.IntegerField;
import com.metaobjects.field.LongField;
import com.metaobjects.field.MetaField;
import com.metaobjects.field.ObjectField;
import com.metaobjects.field.StringField;
import com.metaobjects.object.MetaObject;

import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

/**
 * Pure helper that turns a value-object (VO) {@link MetaObject} into Java source
 * strings for the FR-010 extract codegen:
 *
 * <ul>
 *   <li>{@link #schemaLiteral(MetaObject, String, String)} — a {@code new ExtractSchema(...)}
 *       literal ready for insertion into a generated output-parser class.</li>
 *   <li>{@link #constructorArgs(MetaObject)} — the comma-separated argument list for the
 *       generated {@code new <Payload>(...)} call, reading each component from a
 *       {@code Map<String,Object> d} via the {@code ExtractMap} helpers.</li>
 * </ul>
 *
 * <p>Isolating this logic makes both paths unit-testable without running the full
 * generator pipeline. This class is package-private; the Spring output-parser generator
 * will delegate here.</p>
 *
 * <p>Field-kind mapping (mirrors {@link SpringTypeMapper}):
 * <ul>
 *   <li>{@link EnumField} → {@code FieldKind.ENUM} via {@code FieldSpec.enumField}</li>
 *   <li>{@link StringField} → {@code FieldKind.STRING}</li>
 *   <li>{@link IntegerField} → {@code FieldKind.INT}</li>
 *   <li>{@link LongField} → {@code FieldKind.LONG}</li>
 *   <li>{@link DoubleField} → {@code FieldKind.DOUBLE}</li>
 *   <li>{@link BooleanField} → {@code FieldKind.BOOLEAN}</li>
 *   <li>Nested {@link ObjectField} / non-scalar array → {@code FieldKind.STRING}
 *       with a {@code /* FR-010: nested extract deferred *}{@code /} comment (bounded
 *       deferral).</li>
 * </ul>
 */
final class ExtractSchemaEmitter {

    private ExtractSchemaEmitter() { /* no instances */ }

    // -------------------------------------------------------------------------
    // Public API
    // -------------------------------------------------------------------------

    /**
     * Emit a {@code new ExtractSchema(Format.X, "<rootName>", java.util.List.of(...))}
     * literal for all fields on {@code vo}.
     *
     * @param vo       the payload value-object whose fields drive the schema
     * @param format   {@code "xml"} → {@code Format.XML}; anything else → {@code Format.JSON}
     * @param rootName the logical root name embedded in the schema
     * @return Java source snippet, e.g.
     *         {@code new ExtractSchema(Format.JSON, "Foo", java.util.List.of(...))}
     */
    static String schemaLiteral(MetaObject vo, String format, String rootName) {
        String formatEnum = "xml".equalsIgnoreCase(format) ? "Format.XML" : "Format.JSON";

        List<String> fieldSpecs = new ArrayList<>();
        for (MetaField<?> field : vo.getMetaFields()) {
            fieldSpecs.add(fieldSpecLiteral(field, vo));
        }

        return "new ExtractSchema(" + formatEnum + ", \"" + rootName + "\", java.util.List.of("
            + String.join(", ", fieldSpecs) + "))";
    }

    /**
     * Emit the comma-separated constructor arguments for {@code new <Payload>(...)},
     * reading each field from a {@code Map<String,Object> d} via {@code ExtractMap}.
     *
     * @param vo the payload value-object
     * @return Java source snippet, e.g.
     *         {@code ExtractMap.asString(d, "text"), ExtractMap.asString(d, "confidence")}
     */
    static String constructorArgs(MetaObject vo) {
        List<String> args = new ArrayList<>();
        for (MetaField<?> field : vo.getMetaFields()) {
            args.add(constructorArgForField(field));
        }
        return String.join(", ", args);
    }

    // -------------------------------------------------------------------------
    // Private helpers — schema literal
    // -------------------------------------------------------------------------

    /**
     * Build a {@code FieldSpec.*(...)} call for a single field.
     */
    @SuppressWarnings("rawtypes")
    private static String fieldSpecLiteral(MetaField<?> field, MetaObject owner) {
        String name = field.getName();
        boolean required = isRequired(field);

        if (field instanceof EnumField ef) {
            return enumFieldSpec(name, required, ef, owner);
        }

        // Nested object or array-of-objects: bounded deferral.
        if (field instanceof ObjectField) {
            return "FieldSpec.scalar(\"" + name + "\", FieldKind.STRING, " + required
                + ") /* FR-010: nested extract deferred */";
        }

        // Scalar fallback — same instanceof chain as SpringTypeMapper.
        String kindName = scalarKind(field);
        if (kindName == null) {
            // Unknown field type — fall back to STRING.
            return "FieldSpec.scalar(\"" + name + "\", FieldKind.STRING, " + required
                + ") /* FR-010: unsupported field type, defaulting to STRING */";
        }
        return "FieldSpec.scalar(\"" + name + "\", FieldKind." + kindName + ", " + required + ")";
    }

    /**
     * Build a {@code FieldSpec.enumField(...)} call, including enum values and the
     * alias map (empty map when no {@code @enumAlias} is present).
     */
    private static String enumFieldSpec(String name, boolean required, EnumField field, MetaObject owner) {
        // @values — guaranteed non-null by the loader's ValidationPhase.
        @SuppressWarnings("unchecked")
        List<String> values = (List<String>) field.getMetaAttr(EnumField.ATTR_VALUES).getValue();

        // @enumAlias — optional; produce Map.of() when absent or empty.
        var aliasAttr = field.hasMetaAttr(EnumField.ATTR_ENUM_ALIAS, false)
            ? field.getMetaAttr(EnumField.ATTR_ENUM_ALIAS)
            : null;
        Properties aliases = aliasAttr != null ? (Properties) aliasAttr.getValue() : null;
        String aliasMapLiteral = (aliases != null && !aliases.isEmpty())
            ? buildMapOfLiteral(aliases)
            : "java.util.Map.of()";

        // Build the List.of(...) for enum values.
        List<String> quoted = new ArrayList<>(values.size());
        for (String v : values) quoted.add("\"" + v + "\"");
        String valuesList = "java.util.List.of(" + String.join(", ", quoted) + ")";

        // FR-011: resolve the three new enum args (field → object.value → "strip" for normalize).
        // Keep the back-compat 4-arg form when nothing is set; otherwise emit the 7-arg form
        // (name, required, values, aliases, coerceDefault, normalize, defaultValue).
        String coerceDefault = ownAttrString(field, EnumField.ATTR_COERCE_DEFAULT);
        String defaultValue = ownAttrString(field, EnumField.ATTR_DEFAULT);
        String normalize = resolveNormalize(field, owner);

        if (coerceDefault == null && defaultValue == null && NORMALIZE_DEFAULT.equals(normalize)) {
            return "FieldSpec.enumField(\"" + name + "\", " + required + ", "
                + valuesList + ", " + aliasMapLiteral + ")";
        }

        String cdLit = coerceDefault == null ? "null" : "\"" + escapeJava(coerceDefault) + "\"";
        String normLit = "\"" + normalize + "\"";
        String dvLit = defaultValue == null ? "null" : "\"" + escapeJava(defaultValue) + "\"";

        return "FieldSpec.enumField(\"" + name + "\", " + required + ", "
            + valuesList + ", " + aliasMapLiteral + ", " + cdLit + ", " + normLit + ", " + dvLit + ")";
    }

    /** FR-011 default {@code @normalize} mode ("strip") — mirrors {@link EnumField#NORMALIZE_DEFAULT}. */
    private static final String NORMALIZE_DEFAULT = EnumField.NORMALIZE_DEFAULT;

    /**
     * FR-011: resolve the enum normalization mode for a field — field-level {@code @normalize},
     * else the owning {@code object.value}'s {@code @normalize} (the per-object default), else the
     * global default ("strip"). {@code owner} may be {@code null} (resolution then skips the object
     * tier). Mirrors the TS resolveNormalize and the C# Fr010FieldMapping.ResolveNormalize.
     */
    static String resolveNormalize(MetaField<?> field, MetaObject owner) {
        String fieldMode = ownAttrString(field, EnumField.ATTR_NORMALIZE);
        if (fieldMode != null) return fieldMode;
        String objMode = owner == null ? null : ownAttrString(owner, EnumField.ATTR_NORMALIZE);
        if (objMode != null) return objMode;
        return NORMALIZE_DEFAULT;
    }

    /** The own (non-inherited) string value of an attr on a node, or {@code null} when absent/empty. */
    private static String ownAttrString(com.metaobjects.MetaData node, String attr) {
        if (!node.hasMetaAttr(attr, false)) return null;
        String s = node.getMetaAttr(attr, false).getValueAsString();
        return (s == null || s.isEmpty()) ? null : s;
    }

    /** Escape a value for embedding inside a Java double-quoted string literal. */
    private static String escapeJava(String value) {
        StringBuilder sb = new StringBuilder(value.length() + 4);
        for (int i = 0; i < value.length(); i++) {
            char c = value.charAt(i);
            switch (c) {
                case '\\': sb.append("\\\\"); break;
                case '"':  sb.append("\\\""); break;
                case '\t': sb.append("\\t"); break;
                case '\n': sb.append("\\n"); break;
                case '\r': sb.append("\\r"); break;
                default:   sb.append(c);
            }
        }
        return sb.toString();
    }

    /**
     * Emit a {@code java.util.Map.ofEntries(java.util.Map.entry(k, v), ...)} literal
     * from a {@link Properties}. Entries are sorted by key for deterministic output.
     *
     * <p>{@code java.util.Map.of} only has overloads up to 10 key-value pairs; using
     * {@code Map.ofEntries} removes that arity cap and allows any number of entries.</p>
     */
    private static String buildMapOfLiteral(Properties props) {
        List<String> keys = new ArrayList<>();
        for (Object k : props.keySet()) keys.add(k.toString());
        keys.sort(String::compareTo);

        StringBuilder sb = new StringBuilder("java.util.Map.ofEntries(");
        for (int i = 0; i < keys.size(); i++) {
            if (i > 0) sb.append(", ");
            String k = keys.get(i);
            String v = props.getProperty(k);
            sb.append("java.util.Map.entry(\"").append(k).append("\", \"").append(v).append("\")");
        }
        sb.append(')');
        return sb.toString();
    }

    /**
     * Return the {@code FieldKind} enum member name for a scalar field (matches the
     * {@code instanceof} order in {@link SpringTypeMapper#javaTypeName}), or
     * {@code null} when the field type is not a known scalar.
     */
    @SuppressWarnings("rawtypes")
    private static String scalarKind(MetaField<?> field) {
        if (field instanceof StringField)  return "STRING";
        if (field instanceof IntegerField) return "INT";
        if (field instanceof LongField)    return "LONG";
        if (field instanceof DoubleField)  return "DOUBLE";
        if (field instanceof BooleanField) return "BOOLEAN";
        return null;
    }

    // -------------------------------------------------------------------------
    // Private helpers — constructor args
    // -------------------------------------------------------------------------

    /**
     * Build the {@code ExtractMap.*} call for a single field's constructor argument.
     */
    @SuppressWarnings("rawtypes")
    private static String constructorArgForField(MetaField<?> field) {
        String name = field.getName();

        // Enum fields: string-backed on the wire.
        if (field instanceof EnumField) {
            // enum-array is List<String> on the wire; scalar enum is a String.
            return field.isArray()
                ? "ExtractMap.asStringList(d, \"" + name + "\")"
                : "ExtractMap.asString(d, \"" + name + "\")";
        }

        // Nested object / array-of-objects: deferred in the self-contained path (the
        // record component type is a nested payload or List<NestedPayload>, which the
        // scalar ExtractMap readers can't produce). The runtime-delegating extract
        // overload populates these — see SpringOutputParserGenerator. Emit a typed
        // null so the generated constructor still compiles.
        if (field instanceof ObjectField) {
            return "null /* FR-010: nested extract deferred — use extractLenient(loader, text) */";
        }

        // Scalar array fields: List<String>.
        if (field.isArray()) {
            return "ExtractMap.asStringList(d, \"" + name + "\")";
        }

        // Scalar types.
        if (field instanceof IntegerField) return "ExtractMap.asInt(d, \"" + name + "\")";
        if (field instanceof LongField)    return "ExtractMap.asLong(d, \"" + name + "\")";
        if (field instanceof DoubleField)  return "ExtractMap.asDouble(d, \"" + name + "\")";
        if (field instanceof BooleanField) return "ExtractMap.asBool(d, \"" + name + "\")";

        // StringField (and any other unrecognized type): asString.
        return "ExtractMap.asString(d, \"" + name + "\")";
    }

    // -------------------------------------------------------------------------
    // Private helpers — common
    // -------------------------------------------------------------------------

    /**
     * Returns {@code true} when the field carries {@code @required: true}.
     * Uses {@code getValueAsString()} to handle both the Boolean-attribute and
     * String-attribute storage paths.
     */
    private static boolean isRequired(MetaField<?> field) {
        return field.hasMetaAttr(MetaField.ATTR_REQUIRED)
            && "true".equalsIgnoreCase(
                field.getMetaAttr(MetaField.ATTR_REQUIRED).getValueAsString());
    }
}

File: server/typescript/packages/codegen-ts/src/templates/extract-schema-emitter.ts
// server/typescript/packages/codegen-ts/src/templates/extract-schema-emitter.ts
//
// Turns a payload value-object into TS source fragments for the FR-010 extract codegen:
//   • schemaLiteral     — a `extractSchema(Format.JSON, "root", [ … ])` baked descriptor
//                         built from FieldSpec factories (scalar / enumField).
//   • mirrorInterface   — an all-nullable mirror interface `<Payload>Extracted` (each
//                         component `T | null`); extract returns this nullable twin rather
//                         than the strict payload (same reasoning as the Java/C#/Kotlin ports).
//   • mirrorInitializer — `{ prop: asString(d, "prop"), … }` building the mirror from the
//                         forgiving outcome map `d`.
//
// Mirrors the C# ExtractSchemaEmitter (adapted to TS syntax + the `| null` nullable mirror).
// Bounded scope: scalar / enum / scalar-array. Nested object + array-of-enum deferred.

import {
  type MetaData,
  FIELD_SUBTYPE_ENUM,
  FIELD_SUBTYPE_OBJECT,
  FIELD_ATTR_ENUM_ALIAS,
} from "@metaobjectsdev/metadata";
import {
  fields,
  isRequired,
  isArray,
  xmlText,
  scalarKind,
  mirrorType,
  extractMapCall,
  enumValues,
  coerceDefault,
  defaultValue,
  resolveNormalize,
  jsonStringLiteral,
  stringArrayLiteral,
  propertiesMapLiteral,
} from "./fr010-field-mapping.js";
import { NORMALIZE_DEFAULT } from "@metaobjectsdev/metadata";

/** Emit `extractSchema(Format.X, "rootName", [ … ])`. */
export function schemaLiteral(vo: MetaData, format: string, rootName: string): string {
  const formatEnum = format.toLowerCase() === "xml" ? "Format.XML" : "Format.JSON";
  const specs = fields(vo).map((f) => fieldSpecLiteral(f, vo));
  return `extractSchema(${formatEnum}, ${jsonStringLiteral(rootName)}, [${specs.join(", ")}])`;
}

function fieldSpecLiteral(field: MetaData, owner: MetaData): string {
  const name = jsonStringLiteral(field.name);
  const required = isRequired(field);

  if (field.subType === FIELD_SUBTYPE_ENUM) {
    const valuesLit = stringArrayLiteral(enumValues(field));
    const aliasLit = propertiesMapLiteral(field.ownAttr(FIELD_ATTR_ENUM_ALIAS));
    // FR-011: extended enumField signature is (name, required, values, aliases,
    // coerceDefault?, normalize="strip", defaultValue?). Resolve the three new args
    // (field → object → "strip" for normalize) and emit only what's needed: keep the
    // back-compat 4-arg form when nothing is set, else emit the positional tail up to
    // the last meaningful arg.
    const cd = coerceDefault(field);
    const dv = defaultValue(field);
    const normalize = resolveNormalize(field, owner);
    // enumField() sets array:false; enum-array is a bounded deferral (parity with Java/C#).
    if (cd == null && dv == null && normalize === NORMALIZE_DEFAULT) {
      return `enumField(${name}, ${required}, ${valuesLit}, ${aliasLit})`;
    }
    const cdLit = cd == null ? "null" : jsonStringLiteral(cd);
    const normLit = jsonStringLiteral(normalize);
    if (dv == null) {
      return `enumField(${name}, ${required}, ${valuesLit}, ${aliasLit}, ${cdLit}, ${normLit})`;
    }
    return `enumField(${name}, ${required}, ${valuesLit}, ${aliasLit}, ${cdLit}, ${normLit}, ${jsonStringLiteral(dv)})`;
  }

  if (field.subType === FIELD_SUBTYPE_OBJECT) {
    // FR-010: nested extract deferred — treat as an opaque required/optional string slot.
    return `scalar(${name}, FieldKind.STRING, ${required}) /* FR-010: nested extract deferred */`;
  }

  const kind = scalarKind(field.subType) ?? "STRING";
  // Scalar-array: the scalar() factory only builds singular specs (array:false), so emit a
  // FieldSpec object literal with array:true. Tier-2 win over the Roslyn proof: the emitted
  // extract() actually populates the array at runtime (ExtractMap.asStringList).
  if (isArray(field)) {
    return (
      `{ name: ${name}, kind: FieldKind.${kind}, required: ${required}, array: true, ` +
      `enumValues: null, enumAlias: null, min: null, max: null, nested: null }`
    );
  }
  // @xmlText: a scalar field marked to receive its element's XML text content.
  if (xmlText(field)) {
    return `textContentField(${name}, FieldKind.${kind}, ${required})`;
  }
  return `scalar(${name}, FieldKind.${kind}, ${required})`;
}

/** Emit the all-nullable mirror interface declaration. */
export function mirrorInterface(vo: MetaData, interfaceName: string): string {
  const base = interfaceName.endsWith("Extracted")
    ? interfaceName.slice(0, -"Extracted".length)
    : interfaceName;
  const lines: string[] = [];
  lines.push(
    `/** Best-effort extracted twin of \`${base}\` — every field nullable (null where lost/malformed). */`,
  );
  lines.push(`export interface ${interfaceName} {`);
  for (const f of fields(vo)) {
    lines.push(`  ${f.name}: ${mirrorType(f)};`);
  }
  lines.push("}");
  return lines.join("\n");
}

/** Emit `{ prop: asString(d, "prop"), … }` building the mirror from the forgiving map `d`. */
export function mirrorInitializer(vo: MetaData): string {
  const assigns = fields(vo).map((f) => `${f.name}: ${extractMapCall(f)}`);
  return `{ ${assigns.join(", ")} }`;
}
